proof-systems/src/darlin/fiat-shamir/src/poseidon/constraints.rs

```diff
@@ -0,0 +1,349 @@
+    ) -> Result<(), SynthesisError> {
+        if !self.pending_inputs.is_empty() {
+            // Permute inputs
+            DensityOptimizedPoseidonQuinticSboxHashGadget::<ConstraintF, P, DOP>::apply_inputs_to_state(
+                cs.ns(|| "apply permutations to inputs"),
+                self.pending_inputs.as_slice(),
+                &mut self.state,
+            )?;
+
+            // Clear input buffer
+            self.pending_inputs.clear();
+        } else {
+            // apply permutation to state
+            DensityOptimizedPoseidonQuinticSboxHashGadget::<ConstraintF, P, DOP>::poseidon_perm(
+                cs.ns(|| "permute state"),
+                &mut self.state,
+            )?;
+        }
+
+        // Push RATE many elements from the state into the output buffer
+        self.pending_outputs
+            .extend_from_slice(&self.state[..P::R]);
+
+        Ok(())
+    }
+
+    fn enforce_get_element<CS: ConstraintSystemAbstract<ConstraintF>>(
+        &mut self,
+        cs: CS,
+    ) -> Result<FpGadget<ConstraintF>, SynthesisError> {
+        if self.pending_outputs.is_empty() {
+            self.enforce_permutations(cs)?;
+        }
+
+        Ok(self.pending_outputs.pop().unwrap())
+    }
+
+    /// Enforce squeezing of 'num_bits' Booleans.
+    fn enforce_get_bits<CS: ConstraintSystemAbstract<ConstraintF>>(
+        &mut self,
+        mut cs: CS,
+        num_bits: usize,
+    ) -> Result<Vec<Boolean>, SynthesisError> {
+        if num_bits == 0 {
+            Err(SynthesisError::Other("No challenge to get !".to_string()))?;
+        }
+
+        // We return a certain amount of bits by retrieving field elements instead,
+        // serialize them and return their bits.
+
+        // Smallest number of field elements to retrieve to reach 'num_bits' is ceil(num_bits/FIELD_CAPACITY).
+        // This is done to achieve uniform distribution over the output space, and it also
+        // comes handy as in the circuit we don't need to enforce range proofs for them.
+        let usable_bits = ConstraintF::Params::CAPACITY as usize;
+        let num_elements = (num_bits + usable_bits - 1) / usable_bits;
+        let mut src_elements = Vec::with_capacity(num_elements);
+
+        // Apply as many permutations as needed to get the required number of field elements
+        for i in 0..num_elements {
+            src_elements.push(self.enforce_get_element(cs.ns(|| format!("Get elem {}", i)))?);
+        }
+
+        // Serialize field elements into bits and return them
+        let mut dest_bits: Vec<Boolean> = Vec::with_capacity(usable_bits * num_elements);
+
+        // discard modulus - capacity bits
+        let to_skip = ConstraintF::Params::MODULUS_BITS as usize - usable_bits;
+        for (i, elem) in src_elements.into_iter().enumerate() {
+            let mut elem_bits = elem.to_bits_with_length_restriction(
+                cs.ns(|| format!("elem {} to bits", i)),
+                to_skip,
+            )?;
```

Comment:
We can deal with this in two ways:

- Accept the incompleteness issue, as it happens very seldomly;
- Perform a range proof.

In both cases this will be a temporary solution as when we will have range proofs optimization (costing the same as a to_bits), then we can do range proof without extra cost